### ISSUE TITLE
CA1508 false positive: null check in finally block flagged when null assigned after code that could throw

### ISSUE THREAD

**Issue by jonathanou** [dotnet/roslyn-analyzers#7028]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0-preview.23472.1 (Latest)

### Describe the bug

CA1508 fires incorrectly on code that checks a variable for null in a finally block when the variable is only assigned null after several method/constructor calls, all of which can throw.

### Steps To Reproduce

Add this method and build with CA1508 configured as error.

```cs
public byte[] ReadBytesFromFile()
{
    FileStream? fileStream = null;
    try
    {
        fileStream = File.OpenRead("MyFile.txt");

        using var reader = new BinaryReader(fileStream);

        fileStream = null;
        return reader.ReadBytes(100);
    }
    finally
    {
        if (fileStream != null) // CA1508 incorrectly fires: 'fileStream != null' is always 'false'.
        {
            fileStream.Dispose();
        }
    }
}
```

### Expected behavior

CA1508 does not fire.

### Actual behavior

CA1508 fires on `if(fileStream != null)` in the `finally` block.

## Additional context

Possibly related to #6973

**Comment by mavasani** [comment]
@jonathanou thanks for reporting the issue, I am able to repro it on the latest analyzer bits. It seems `using var` statement is causing the analyzer to report the false positive, removing the `using` leads to no false positive.

**Comment by jonathanou** [comment]
@mavasani , thanks for verifying the issue!  I didn't realize `using` was causing the false-positive, I've updated the title and description.
